File: backend/app/services/anomaly.py
"""Anomaly score: statistical deviation (z-score) vs. the recent baseline.

Compares the current CPU, memory and load readings against their behavior
over the last N minutes. A score of 0 means "behaves as usual"; higher
scores mean the server is behaving unusually for its own baseline — the
idea popularized by Netdata, implemented here with plain statistics.
"""

import math
from statistics import mean, pstdev
from typing import Optional


def _z_score(value: float, baseline: list[float]) -> float:
    if len(baseline) < 2:
        return 0.0
    average = mean(baseline)
    deviation = pstdev(baseline)
    if deviation < 1e-9:
        return 0.0
    return max(0.0, (value - average) / deviation)


def _clamp_to_hundred(value: float) -> float:
    return max(0.0, min(100.0, value / 4.0 * 100.0))


def _level(score: float) -> str:
    if score < 40:
        return "normal"
    if score < 60:
        return "elevated"
    if score < 80:
        return "high"
    return "critical"


class AnomalyScorer:
    def __init__(self, window_minutes: int = 120, critical: float = 80.0):
        self.window_minutes = window_minutes
        self.critical = critical

    def score(
        self,
        cpu_percent: float,
        memory_percent: float,
        load_one_min: Optional[float],
        history_points: list[dict],
    ) -> dict:
        """Returns the anomaly report for the current sample."""
        cpu_baseline = [p.get("cpu_percent", 0) for p in history_points]
        mem_baseline = [p.get("memory_percent", 0) for p in history_points]
        load_baseline = [p.get("load_one_min") for p in history_points if p.get("load_one_min") is not None]

        metrics: dict[str, float] = {"cpu": _clamp_to_hundred(_z_score(cpu_percent, cpu_baseline))}
        metrics["memory"] = _clamp_to_hundred(_z_score(memory_percent, mem_baseline))
        if load_one_min is not None and load_baseline:
            metrics["load"] = _clamp_to_hundred(_z_score(load_one_min, load_baseline))

        weights = {"cpu": 0.4, "memory": 0.3, "load": 0.3}
        weight_sum = sum(weights[k] for k in metrics)
        score = (
            sum(weights[k] * metrics[k] for k in metrics) / weight_sum
            if weight_sum > 0
            else 0.0
        )
        return {
            "score": round(score, 1),
            "level": _level(score),
            "critical_threshold": self.critical,
            "metrics": {k: round(v, 1) for k, v in metrics.items()},
        }

File: backend-node/src/metrics/anomaly.ts
interface HistoryLike {
  cpu_percent?: number;
  memory_percent?: number;
  load_one_min?: number | null;
}

function mean(values: number[]): number {
  return values.reduce((sum, v) => sum + v, 0) / values.length;
}

function pstdev(values: number[]): number {
  const n = values.length;
  if (n < 2) return 0;
  const avg = mean(values);
  const variance = values.reduce((sum, v) => sum + (v - avg) ** 2, 0) / n;
  return Math.sqrt(variance);
}

function zScore(value: number, baseline: number[]): number {
  if (baseline.length < 2) return 0;
  const avg = mean(baseline);
  const deviation = pstdev(baseline);
  if (deviation < 1e-9) return 0;
  return Math.max(0, (value - avg) / deviation);
}

function clampToHundred(value: number): number {
  return Math.max(0, Math.min(100, value / 4 * 100));
}

function level(score: number): string {
  if (score < 40) return 'normal';
  if (score < 60) return 'elevated';
  if (score < 80) return 'high';
  return 'critical';
}

export class AnomalyScorer {
  constructor(
    private readonly windowMinutes: number = 120,
    private readonly critical: number = 80,
  ) {}

  score(
    cpuPercent: number,
    memoryPercent: number,
    loadOneMin: number | null,
    historyPoints: HistoryLike[],
  ): { score: number; level: string; critical_threshold: number; metrics: Record<string, number> } {
    const cpuBaseline = historyPoints.map((p) => p.cpu_percent ?? 0);
    const memBaseline = historyPoints.map((p) => p.memory_percent ?? 0);
    const loadBaseline = historyPoints
      .map((p) => p.load_one_min)
      .filter((v): v is number => v != null);

    const metrics: Record<string, number> = {
      cpu: clampToHundred(zScore(cpuPercent, cpuBaseline)),
      memory: clampToHundred(zScore(memoryPercent, memBaseline)),
    };
    if (loadOneMin != null && loadBaseline.length > 0) {
      metrics['load'] = clampToHundred(zScore(loadOneMin, loadBaseline));
    }

    const weights: Record<string, number> = { cpu: 0.4, memory: 0.3, load: 0.3 };
    const weightSum = Object.keys(metrics).reduce(
      (sum, key) => sum + (weights[key] ?? 0),
      0,
    );
    const score =
      weightSum > 0
        ? Object.keys(metrics).reduce(
            (sum, key) => sum + (weights[key] ?? 0) * metrics[key],
            0,
          ) / weightSum
        : 0;

    const roundedMetrics: Record<string, number> = {};
    for (const key of Object.keys(metrics)) {
      roundedMetrics[key] = Math.round(metrics[key] * 10) / 10;
    }

    return {
      score: Math.round(score * 10) / 10,
      level: level(score),
      critical_threshold: this.critical,
      metrics: roundedMetrics,
    };
  }
}